No container locks on `docker ps`

fabiokung opened the pull request:
**- What I did**

Fixes #28183

We are seeing a fair amount of lockups on `docker ps`. More details [here](https://github.com/docker/docker/issues/28183#issuecomment-278020462).

While things seem to get better with every new version of Docker, there may always be a legitimate reason to hold container locks during some not-very-quick operations. Container queries (`docker ps`) currently try to grab a lock on every single container being inspected. The risk for lockups is big.

I started by trying to pick up @cpuguy83's work on #30225 (the [memdb branch](https://github.com/docker/docker/compare/master...cpuguy83:container_memdb)). Completely moving the container in-memory store to a consistent ACID DB is a noble cause, but I quickly realized it would be a huge undertaking (references that would need to be deep copied, structs that would need to be broken apart, etc.).

This is a more incremental step towards that goal. Containers keep being stored in the existing in-memory store, and mutations still grab a `Lock()` to avoid races. We then keep a consistent view of all containers rendered during all of these mutations, so readers (queries) do not need to `Lock()` anything.

Queries and `docker ps` are now very cheap. There are now virtually no chances of them getting stuck in a lockup, at the expense of some more lock contention during some mutation operations, because the current `MemDB` implementation (using hashicorp/go-memdb) does a table-level `Lock()` during write transactions.

These write locks are very short and hopefully won't be a problem (`memdb.Save()`). **If they are**, in the future another in-memory ACID implementation that supports row level locking could be investigated. Or replications could be done asynchronously (and optimistically), reducing lock contentions but causing the read-only view to be eventually consistent.

There is also admittedly some risk of missing parts of the code that are mutating containers and not replicating state when they should. However, we already replicate container state to disk (`container.ToDisk()`), then any of these cases should be treated as bugs and covered by tests.

**- How I did it**

All data that is necessary to serve queries (`docker ps`) is snapshotted during operations that mutate that data. Typically these mutations already hold a lock on the container object they are mutating. All places in the code calling `container.ToDisk()` and `container.ToDiskLocking()` are good candidates to also replicate state to the in-memory rendered consistent view.

Queries use a read-only transaction on the replicated in-memory DB, and don't need to grab locks on each individual container being inspected.

In the future, more read operations can be served from the in-memory ACID store, e.g.:

* `docker inspect` (shouldn't be too hard, but will require deep copying some pointers currently being held by the container during `Snapshot()`)
* healthcheck probe results
* stats collection
* `container.RWLayer`
* `ExecConfigs`

This way we can incrementally move things as needed/wanted and even potentially one day completely eliminating container locks (#30225). One day `container.ToDisk()` could also be replaced by checkpointing the in-memory DB to disk.

I explicitly left out some code paths that mutate parts of the container that queries don't currently care about:

* `container.RemovalInProgress`
* `container.BaseFS` (all calls to `mount.Mount()` may mutate the container: `daemon.ContainerArchivePath`, `daemon.ContainerCopy`, `daemon.ContainerExtractToDir`, daemon.`ContainerStatPath`, ... )
* `container.HasBeenManuallyStopped`
* `container.HostsPath`, `container.ResolvConfPath`: being mutated by some networking code paths

**- How to verify it**

All related `docker ps` and `GetContainerApi` cli-integration tests are passing. This should be treated as an internal refactoring and no functionality is being added or changed.

**- Description for the changelog**

lock-free docker ps, reducing the chances of daemon lockups during queries

**- A picture of a cute animal (not mandatory but encouraged)**

My dog, Jilly:

![my dog: Jilly](https://pbs.twimg.com/media/BxrUBhlCUAA9LTe.jpg:medium)

fabiokung committed:
Grab a lock to read container.RemovalInProgress

Signed-off-by: Fabio Kung <fabio.kung@gmail.com>

fabiokung committed:
Move platform specific mount data to Container

The Solaris version (previously daemon/inspect_solaris.go) was
apparently missing some fields that should be available on that
platform.

Signed-off-by: Fabio Kung <fabio.kung@gmail.com>

fabiokung committed:
in-memory ACID store for containers

This can be used by readers/queries so they don't need locks.

Signed-off-by: Fabio Kung <fabio.kung@gmail.com>

fabiokung committed:
keep a consistent view of containers rendered

Replicate relevant mutations to the in-memory ACID store. Readers will
then be able to query container state without locking.

Signed-off-by: Fabio Kung <fabio.kung@gmail.com>

fabiokung committed:
stop grabbing container locks during ps

Container queries are now served from the consistent in-memory db, and
don't need to grab a lock on every container being listed.

Signed-off-by: Fabio Kung <fabio.kung@gmail.com>

icecrime commented:
Ping @cpuguy83 @tonistiigi 🎉

fabiokung commented:
also related to #28754

thaJeztah commented:
Thanks for working on this @fabiokung !

tonistiigi commented on container/snapshot.go:
```diff
@@ -0,0 +1,154 @@
+		StartedAt:    container.StartedAt,
+		Running:      container.Running,
+		Paused:       container.Paused,
+		Restarted:    container.Paused,
```
Restarted

fabiokung replied:
oops, good catch

aaronlehmann commented:
I like the idea. This looks way simpler than the original iteration that involved deep copies.

It looks like `daemon.containersReplica.Save(c.Snapshot())` is a common pattern, and it needs to be protected by the container lock to avoid transactions overwriting each other. Maybe it would be good idea to define a method on `Container` that does this, and add a prominent comment that it should only be called with the lock held (or it could acquire the lock itself, if this pattern works for the call sites).

fabiokung commented:
Good point on moving the replication to `Container`. I'll take a stab at it.

tonistiigi commented:
I'm bit torn about this. I stand by my earlier comment that we shouldn't be afraid of locks, just the wrong implementation of them. This problem doesn't only show in performance, most places we call `IsRunning/IsPaused/IsRestarting` are clear races. If we had proper synchronization between lifecycle states, a plain `RWMutex` should probably even perform better than this without the extra maintenance cost. But I don't see anyone taking on this refactor, at least before everything about containerd integration is cleared. This version does look simpler than the earlier one. Not going to block if other maintainers want to go forward with it.

cpuguy83 commented on container/snapshot.go:
```diff
@@ -0,0 +1,152 @@
+}
+
+// Snapshot provides a read only view of a Container. Callers must hold a Lock on the container object.
+func (container *Container) Snapshot() *Snapshot {
```
This is effectively the same thing as the deep copy, and has to be maintained in the same way.
It ends up being pretty fragile as any change even to any of the underlying types could wind up messing things up... e.g. adding a new reference type to a struct which is not properly copied.

fabiokung replied:
It is similar, but not necessarily the same. It's more of a mapping to what queries need to see, rather than only a (deep) copy. The rest of the code already has precedence for similar mapping to different views/strucs being done at different layers.

It is fragile in the same sense that public API handlers can break when new fields are added and not properly mapped to strucs being serialized for responses.

cpuguy83 replied:
Instead of defining a new type do you think we could use the API container type?

fabiokung replied:
I started with that, but it had a pointer to `*NetworkSettings` and was missing a few fields that `daemon/list.go` requires. Since `types.Container` is a public API, I didn't want break API compatibility or mess with its JSON serialization.

It's probably possible, but a bit riskier I think.

fabiokung commented:
@tonistiigi fair point. This doesn't block going back to locking for reads when/if you are comfortable enough that all concurrency and locking is being properly handled across the codebase. It would be a bit of throwaway work, but at least it fixes the problem we are having today in the short term, until more long term solutions (what you propose with proper synchronization between lifecycle states, or "lock-free" transactional mutations everywhere, ...).

I'm not proposing we lock down on a decision on how concurrency will be handled long term, but this is a big pain right now and I don't feel we can wait much for a big re-architecture to happen.

fabiokung committed:
Move checkpointing to the Container object

Also hide ViewDB behind an inteface.

Signed-off-by: Fabio Kung <fabio.kung@gmail.com>

fabiokung commented:
@aaronlehmann I moved the checkpointing/replication operation to `*Container`. How does it look now?

I liked it more, since now it is clearer that checkpointing should probably be done when saving state to disk. It also allowed `container.snapshot()` to not be public.

aaronlehmann commented:
Thanks, I think that's an improvement.

dnephin commented:
This needs a rebase, but it looks like some of the requested changes were made, so could use another review as well

fabiokung commented:
Rebased.

fabiokung committed:
Make tests a bit less flaky

Prevent tests failing when teardown tries to delete a container that is
already being removed.

Signed-off-by: Fabio Kung <fabio.kung@gmail.com>

AkihiroSuda commented on container/snapshot.go:
```diff
@@ -0,0 +1,152 @@
+	"github.com/docker/go-connections/nat"
+)
+
+// Snapshot is a read only view for Containers
```
Please add some comment about how to maintain this structure?
e.g. Why `HostConfig.{NetworkMode, Isolation}` is needed but others not?

AkihiroSuda replied:
Also, I wonder we can embed `api/types.Container` for deduplication

fabiokung replied:
I'll add a comment about the structure.

Re: using `types.Container`, here's a brief discussion about it we had on a commit that was previously on this PR: https://github.com/docker/docker/pull/31273#discussion_r102825454

cpuguy83 commented:
I wonder if instead of having this intermediate object, we can store a duplicate of the container object like so:

(warning, pseudo-code follows)
```go
func(c *Container) ToDisk() {
   // normal stuff

 snapshot := &Container{ID: c.ID}
 snapshot.FromDisk()
  c.snapshotsStore.Add(snapshot)
}
```

fabiokung commented:
@cpuguy83 isn't it what we discussed here: https://github.com/docker/docker/pull/31273#discussion_r102825454 ?

Or is it something else I'm missing?

fabiokung committed:
how to maintain the container snapshot struct

Signed-off-by: Fabio Kung <fabio.kung@gmail.com>

cpuguy83 commented:
@fabiokung Not the same, this would be the actual `*container.Container` type which gets stored in the snapshot which is fully-copied by definition since it'll be unmarshalled from the on-disk json.

fabiokung commented:
@cpuguy83 ah, I see, I misunderstood.

It seems like a matter of taste or what you want to optimize for with this. Keeping the `Snapshot` struct means it will be easier to make durable persistence be done off the ACID in-memory store someday. i.e.: replace `container.ToDisk()` with `containerstore.SnapshotToDisk()`, or even make it transparent by using a persistent implementation of the ACID container store (sqllite3, leveldb, etc).

Feeding the in-memory store from the current json serialized files means a bit less duplication, but harder to move away from that serialization strategy. It would also mean putting objects with pointers in the store, which is a bit riskier since any code path can start modifying state kept by the store and break its consistency guarantees.

fabiokung commented:
... and my personal opinion is that risking breaking the consistency of the in-memory store is not worth the small amount of deduplication.

cpuguy83 commented:
@fabiokung The snapshot object you created is storing reference types as it is, not to mention the real container object itself can be written to w/o proper locking and break consistency and potentially even crash the daemon.

I'm not sure I'm following how this intermediate object helps move us towards a different solution. The on-disk representation is the snapshot just not particularly optimized for random access like we'd get from a true database.

fabiokung commented:
@cpuguy83 did you make your mind, or still thinking about it? This is a simple change I think, and I'd rather get this merged than try and force a strong opinion.

cpuguy83 commented:
I would prefer to simplest change for this intermediate solution which is to snapshot from unmarshalled json.
Later we can change `ToDisk` to actually go to something like bolt.

@tonistiigi @aaronlehmann

aaronlehmann commented:
@cpuguy83: Is the idea to make a deep copy by marshalling to json and unmarshalling back into another object?

cpuguy83 commented:
@aaronlehmann Yes, particularly we're already marshalling it, it would just be unmarshalling back into a new object.

aaronlehmann commented:
I guess. It feels less than ideal to rely on reflection for this copying, but if we're already marshalling the object for other reasons, it's (hopefully) not happening in any fast paths.

Avoiding hard-to-maintain code that creates a copy of the container for the store would be good.

There's still the issue of the store not being able to make copies on demand to return to callers. However, I don't think the current PR is doing this completely either. I notice that the struct it's returning has some slices, and the contents of these won't be copied by a struct copy.

Could we define an interface that provides read-only access to the applicable `Container` fields, and return this interface type from the store instead of an actual `Container` struct? Then it would be way harder to inadvertently make changes to an object in the store. This would also save the overhead of a deep copy on every access.

fabiokung commented:
@aaronlehmann do you think an interface with multiple acessors that will copy individual values is much different from a single method that returns a copied struct? I.e.:

```go
type ContainerView interface {
	GetCommand() string
	GetMounts() []types.MountPoint
	// ...
}

func (c *Container) GetCommand() string {
	return c.Command
}

func (c *Container) GetMounts []types.MountPoint {
	return deepCopy(c.Mounts)
}
```

vs (assuming that the `Snapshot` type stops holding references:

```go
func (c *Container) Snapshot() Snapshot {
	return Snapshot{
		Command: c.Command,
		Mounts: deepCopy(c.Mounts),
		// ...
	}
}
```

aaronlehmann commented:
Yeah, it's a bit different. One difference is that most accessors don't need to make copies at all. It seems a bit wasteful to copy, say, `c.Mounts`, if the code that is fetching the container from the store is not going to look at mounts. Another difference is that if a new field is added to `Container` but the interface isn't updated, the accessor method simply won't exist. This is better than code encountering an empty field and getting no compile-time indication that something is wrong.

The interface approach may be impractical if every field needs to exposed. I had the impression that the code doing the fast view-only accesses through the store would only need a subset of the fields. If they need all the fields, maybe it doesn't make sense.

cpuguy83 commented:
To start it could just be interface methods to convert to API types.

fabiokung commented:
yeah, I kinda like that. I'll give it a try.

fabiokung committed:
no need to save state to disk here

State will be saved on the following operation once the container is
properly registered on the daemon.

Signed-off-by: Fabio Kung <fabio.kung@gmail.com>

fabiokung committed:
verifyVolumesInfo needs a container lock

It operates on containers that have already been registered on the
daemon, and are visible to other goroutines.

Signed-off-by: Fabio Kung <fabio.kung@gmail.com>

fabiokung committed:
no need to save container state here

it is already being saved (with a lock held) on the subsequent
operations.

Signed-off-by: Fabio Kung <fabio.kung@gmail.com>

fabiokung committed:
save deep copies of Container in the replica store

Reuse existing structures and rely on json serialization to deep copy
Container objects.

Also consolidate all "save" operations on container.CheckpointTo, which
now both saves a serialized json to disk, and replicates state to the
ACID in-memory store.

Signed-off-by: Fabio Kung <fabio.kung@gmail.com>

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "consistent-ro-view" git@github.com:fabiokung/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842354043080
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

fabiokung commented:
@cpuguy83 how does it look now?

The in-memory store is now holding (deep) copies `FromDisk`, and the transformation to API types happen inside the store layer. I decided to keep the `Snapshot` API since it includes extra fields that queries need to filter on.

I also consolidated all container persistence calls in `container.CheckpointTo`.

cpuguy83 commented:
These seems relatively sane.
SGTM

cpuguy83 commented:
@aaronlehmann @tonistiigi WDYT?

aaronlehmann commented on container/view.go:
```diff
@@ -0,0 +1,288 @@
+		return false, nil,
+			fmt.Errorf("embedded struct '%s' for %#v is invalid", e.EmbeddedStruct, obj)
+	}
+	return e.Field.FromObject(ev.Interface())
```
Why is reflection necessary here? It looks like this is only used to index by ID. Couldn't this function be written to just return the ID from a `Snapshot` struct? (see for example https://github.com/docker/swarmkit/blob/bump_v1.13.2/manager/state/store/clusters.go#L192-L211)

fabiokung replied:
ah, ofc! much better.

aaronlehmann commented on container/view.go:
```diff
@@ -0,0 +1,288 @@
+	txn := db.store.Txn(true)
+	defer txn.Commit()
+	deepCopy := NewBaseContainer(c.ID, c.Root)
+	err := deepCopy.FromDisk() // TODO: deal with reserveLabel
```
`Save` is always called right after `toDisk`, so it seems like a waste to reopen the file and reread it from the disk. Can the caller pass in the JSON instead requiring `Save` to reread it?

fabiokung replied:
the tradeoff here is a less clean API for the store, and having to hold (buffer) the whole json contents in memory, which is not being done currently (the JSON encoder writes directly to the file).

Are you sure it would be better? I'm inclined to say it is not worth it.

aaronlehmann replied:
It seems way better to me. Right now we're passing in a `Container` to `Save`, but that container isn't really used, except for the ID to find the file on disk. I think it would be clearer to pass in the JSON instead. I also think holding the JSON contents in memory between `toDisk` and `Save` is a good thing, not a downside. Opening the file to reread it on the disk when we had it in memory a few microseconds ago doesn't make sense to me.

fabiokung replied:
I gave it a quick try, there are two JSONs per container (`HostConfig` gets saved to a different file). It started to look pretty ugly tbh. I'm really unsure if this is worth it. What's the concern with reopening these files and rereading them? These files are likely to be in the fs cache anyway.

aaronlehmann replied:
Mostly concerned about hurting performance. Reading from the disk isn't necessarily that cheap - it involves a few syscalls, and on filesystems with `atime` enabled, it can actually trigger a disk write to update the metadata. It just didn't make sense to be reading the JSON from disk when we're generating the JSON literally one line up. I guess if it really makes things cleaner and it doesn't cause noticeable performance issues, I'll stop making a big deal out of it, but it did seemed like a strange choice.

fabiokung replied:
@aaronlehmann alright, I made some changes to avoid re-reading the json files. How does it look now?

aaronlehmann replied:
Thanks

aaronlehmann commented on container/view.go:
```diff
@@ -0,0 +1,288 @@
+}
+
+// transform maps a (deep) copied Container object to what queries need
+func (v *memdbView) transform(container *Container) *Snapshot {
```
I was initially put off by this, but I realize now that this is just moving the existing `transformContainer` function (which converts containers to the API type) into the store. I guess that's okay. It sort of makes sense to avoid exposing containers from the store in the native `Container` type, so they don't get used unintentionally in code paths that expect a live `Container` object. On the other hand, having three types to represent a container starts to get weird (`container.Container, `types.Container`, `container.Snapshot`). I wonder if we should return un-transformed containers and make conversion to the API `types.Container` type a separate responsibility. It would at least eliminate one of these three types, if I understand correctly.

fabiokung replied:
I had it returning the original un-transformed containers, but as you pointed out, not exposing them has the benefit that they don't get unintentionally used on code paths that expect a real `Container` object. Also, it would expose a mutable `container.Container` object that's being referenced by the store and can be mutated (unintentionally), breaking the consistency guarantees of the store.

fabiokung committed:
only Daemon.load needs to call label.ReserveLabel

Signed-off-by: Fabio Kung <fabio.kung@gmail.com>

fabiokung committed:
avoid re-reading json files when copying containers

Signed-off-by: Fabio Kung <fabio.kung@gmail.com>

aaronlehmann commented:
Design LGTM

One of the things I notice reading over this code is that there are a lot of places where exported method (such as `CheckpointTo`) are called with the expectation that a lock is already held. This is unconventional and could lead to bugs later. It looks like it's a longstanding issue that's not new with this PR, thus fixing it is probably out of scope, but it's an idea for something to improve in the future.

fabiokung committed:
ensure heath monitor status updates are propagated

initHealthMonitor and updateHealthMonitor can cause container state to
be changed (State.Health).

Signed-off-by: Fabio Kung <fabio.kung@gmail.com>

aaronlehmann commented:
@fabiokung: I see that some integration tests are failing. Could you please take a look at those?

fabiokung commented:
@aaronlehmann on it. I fixed most of it already (sorting issues solved by bringing back nanosec resolution on `CreatedAt` and `unhealthy` states while the monitor is being shutdown). There is a last one failing that I'm having trouble running locally (`DockerDaemonSuite.TestDaemonBackcompatPre17Volumes`), but I'll try to fix it today.

fabiokung committed:
delete unused code (daemon.Start)

Signed-off-by: Fabio Kung <fabio.kung@gmail.com>

fabiokung committed:
avoid saving container state to disk before daemon.Register

Migrate legacy volumes (Daemon.verifyVolumesInfo) before containers are
registered on the Daemon, so state on disk is not overwritten and legacy
fields lost during registration.

Signed-off-by: Fabio Kung <fabio.kung@gmail.com>

fabiokung commented:
All green!

/cc @cpuguy83 @aaronlehmann @tonistiigi

aaronlehmann commented on container/view.go:
```diff
@@ -0,0 +1,287 @@
+func (e *containerByIDIndexer) FromObject(obj interface{}) (bool, []byte, error) {
+	c, ok := obj.(*Container)
+	if !ok {
+		return false, nil, fmt.Errorf("%+v is not a Container", obj)
```
I'd suggest using `%T` to print the type instead of dumping the object value.

aaronlehmann commented on integration-cli/environment/clean.go:
```diff
@@ -50,14 +51,19 @@ func getPausedContainers(t testingT, dockerBinary string) []string {
+			// We can safely ignore those.
+			if strings.Contains(
+				result.Stderr(), "No such container",
+			) || alreadyExists.MatchString(result.Stderr()) {
```
The formatting is a bit weird here. I would move everything up to and including the `||` onto the first line.

fabiokung replied:
I will make the change. I broke the line because the rest of the codebase seemed to be under(-ish) 120 columns.

aaronlehmann commented:
@tonistiigi @cpuguy83 Is this ready to move to code review?

fabiokung commented:
ping, anything else needed here?

cpuguy83 commented:
Design OK.

cpuguy83 commented:
Moving to code-review so we can move forward.

fabiokung commented on container/container.go:
```diff
@@ -132,41 +131,51 @@ func (container *Container) FromDisk() error {
 
-	enc := json.NewEncoder(jsonSource)
+	w := io.MultiWriter(&buf, jsonSource)
+	if err := json.NewEncoder(w).Encode(container); err != nil {
```
any good reason not to? I only did it this way because it seemed cleaner, but it was totally a matter of preference.

thaJeztah commented:
ping @aaronlehmann @AkihiroSuda PTAL

cpuguy83 commented:
I've looked through this a few times, looks good.
Will put this through some concurrency tests as well.

AkihiroSuda commented on container/view.go:
```diff
@@ -0,0 +1,287 @@
+}
+
+// transform maps a (deep) copied Container object to what queries need
+func (v *memdbView) transform(container *Container) *Snapshot {
```
> func (v *memdbView) transform(container *Container) *Snapshot 

I think this function should be non-receiver (i.e. decoupled from `v`) so that we can easily add UTs


WDYT? @cpuguy83

cpuguy83 replied:
Yep, since this isn't really doing anything in the db it can be decoupled.

fabiokung replied:
It depends on a `nameIndex` maintained by the store though.

AkihiroSuda commented:
@fabiokung @cpuguy83 
What's current status?

cpuguy83 commented:
Ok, tested this out, grabbed some metrics from it, all LGTM.
It does need a rebase though (ugh, sorry). @fabiokung Is this something you want to do, or I can carry if needed.

fabiokung commented:
@cpuguy83 thanks! Let me take a quick look and see how bad the conflicts are.

fabiokung commented:
@cpuguy83 Rebased and addressed the remaining feedback from reviews.

Feel free to make any necessary changes that come out of test runs too, if you feel so inclined.

fabiokung committed:
Net operations already hold locks to containers

Fix a deadlock caused by re-entrant locks on container objects.

Signed-off-by: Fabio Kung <fabio.kung@gmail.com>

aaronlehmann commented on container/view_test.go:
```diff
@@ -0,0 +1,106 @@
+	}
+	if err := db.Delete(c); err != nil {
+		t.Fatal(err)
+	}
```
Some of these tests might benefit from using `testify`... i.e. `assert.NoError(t, err)` instead of `if` statements. But the current approach is fine as well.

fabiokung replied:
I'm personally not a big fan, but I'm happy to do what maintainers prefer to keep the codebase consistent.

fabiokung replied:
It seems like all other tests in the same package are using the vanilla `"testing"`. Let's leave it as-is then, and make changes as needed in followup PRs?

aaronlehmann commented on daemon/delete.go:
```diff
@@ -103,14 +103,20 @@ func (daemon *Daemon) cleanupContainer(container *container.Container, forceRemo
 	// container meta file got removed from disk, then a restart of
 	// docker should not make a dead container alive.
-	if err := container.ToDiskLocking(); err != nil && !os.IsNotExist(err) {
+	if err := container.CheckpointTo(daemon.containersReplica); err != nil && !os.IsNotExist(err) {
```
Why are we checkpointing the same container twice in a row? Is this a rebase artifact, perhaps?

fabiokung replied:
good catch!

aaronlehmann commented on daemon/list.go:
```diff
@@ -544,106 +559,21 @@ func includeContainerInList(container *container.Container, ctx *listContext) it
-
-	image := container.Config.Image // if possible keep the original ref
-	if image != container.ImageID.String() {
+// refreshImage when possible and make sure Image is an ID
```
The comment isn't clear.

aaronlehmann commented on daemon/monitor.go:
```diff
@@ -90,7 +93,7 @@ func (daemon *Daemon) StateChanged(id string, e libcontainerd.StateInfo) error {
 		daemon.setStateCounter(c)
 
 		defer c.Unlock()
```
Unrelated to these changes, but it's odd that this line isn't up near the call to `Lock`.

fabiokung commented on container/container_windows.go:
```diff
@@ -125,11 +126,8 @@ func (container *Container) TmpfsMounts() ([]Mount, error) {
 }
 
-// UpdateContainer updates configuration of a container
+// UpdateContainer updates configuration of a container. Callers must hold a Lock on the Container.
```
The places where this is called are holding locks in broader contexts (for more operations), it made sense to let callers decide when to lock.

fabiokung commented:
Test failures for windows seems to be unrelated. Some flakiness with networking on the jenkins slaves.

fabiokung commented:
@aaronlehmann I believe I addressed all your comments. Thanks for reviewing!

aaronlehmann commented:
LGTM

cpuguy83 reviewed (approved):
LGTM

thaJeztah commented:
Whoop! Thank you so much for your contribution @fabiokung! Really happy to see this merged, and hope this will resolve the issues!

gyliu513 commented:
Does this only available in the latest version? Any plan to back merge this?

AkihiroSuda commented:
This is very major change and unlikely to be backported

thaJeztah commented:
Correct, this won't be back ported - there's too much risk in back porting such a big change (which can easily lead to introducing regressions).

This change will be included in the upcoming 17.07 release of Docker CE

gyliu513 commented:
Is it possible to provide a workaround patch for old versions? Many customers often hit `docker ps` hang due to a problematic container. In this condition, I cannot kill the problematic container, the docker daemon also cannot be started, but I have to restart my docker server. Sometimes it is not acceptable to restart a host in a production environment. @thaJeztah

thaJeztah commented:
@gyliu513 there's no "one size fits all" workaround. As suggested in your other issue, I'd first recommend upgrading to a more recent version of docker. If the issue still remains, try to narrow down what actually _causes_ the deadlock. Even though deadlocks can happen, they should really not be frequent. When hitting this situation, please open an issue, and send a `SIGUSR1` to the docker daemon. This should create a stack dump in the logs, which may help investigating what's causing the deadlock.

gyliu513 commented:
Thanks @thaJeztah , but the problem is with current 1.6.2, when I run `sudo kill -SIGUSR1 <docker pid>`, there is no output...

thaJeztah commented:
The dump shows in the daemon logs, not on the command-iine. I'm not sure docker 1.6.2 had this already; possibly it needs `debug` to be enabled on the daemon

thaJeztah commented:
That debugging functionality was added in docker 1.7 https://github.com/moby/moby/pull/10786

Anyway, this is not really the best location for this discussion - feel free to message me on Slack if you need assistance

gyliu513 commented:
I googled docker slack channel before, but no help info, what is the link of docker slack channel? @thaJeztah

thaJeztah commented:
Here's more info about the slack channel; https://blog.docker.com/2016/11/introducing-docker-community-directory-docker-community-slack/, register here for the docker community, and to get an invitation for the Slack channel: http://dockr.ly/community

gyliu513 commented:
Thanks @thaJeztah , as I cannot be approved to the community, I want to ask you the last question here, hope it is OK. ;-)

Regarding the `debug` mode, it is really helpful for troubleshooting. The question is: is it good to enable the `debug` mode in a production env?

gyliu513 commented:
Got the answer myself, we can [turn on the `debug` mode automatically](https://success.docker.com/KBase/How_do_I_enable_%22debug%22_logging_of_the_Docker_daemon%3F), cool!

gyliu513 commented:
@fabiokung 

> Container queries (docker ps) currently try to grab a lock on every single container being inspected

One question for the issue, before your fix, why does `docker ps` need to grab a lock for each single container? Why read-only operations require a lock?

fabiokung commented:
@gyliu513 it uses locks to prevent partial reads and data corruption, since container data is being concurrently modified. Even in read-only operations, locking is how is ensures the state being read will be consistent.

This fix only applies to `docker ps`, and was a first step in fixing others you mentioned. I did mention them in the description. `docker inspect` will also hang, if you happen to hit one of the containers that has its lock held somewhere else.

krmayankk commented:
@fabiokung this fix is available in what docker versions ?

thaJeztah commented:
@krnayankk this will be included in docker 17.07 and up

thiagoalves commented:
@thaJeztah How about docker-ee? Will we have it in 17.06-ee3 or ee4? Or should we wait until 17.09?

cpuguy83 commented:
@thiagoalves It will not be in EE until the next major release. I do not see this one getting backported.

cpuguy83 commented:
@thiagoalves btw, if you have a specific case where you are seeing a deadlock, please report so it can be fixed. This patch is just making the deadlock less apparent, not actually fixing deadlocks.

thiagoalves commented:
@mlespiau

robertglen commented:
17.07 sounds great,  when can we expect to see it?   I'm only seeing 17.05 with the last commit from May of this year.  and oldschool docker 1.13.1 from Feb of this year.  

How long do critical fixes affecting an entire ecosystem of projects typically bake before a release is cut?

cpuguy83 commented:
@robertglen 17.07 was released in July....

tonistiigi commented:
@robertglen You probably need to update your apt/yum repositories from dockerproject.org to download.docker.com. https://docs.docker.com/engine/installation/

robertglen commented:
heh, OK I just went to the code portion of this repo and looked at branches and tags and it stopped at 17.05 and had completely dismissed visiting docker's website :\  my bad.

euank commented:
@robertglen 

The Moby project doesn't make releases of this repo, rather Docker Inc makes releases of the Docker Community Edition software (which includes code from this repo). Those releases are tagged in the [docker-ce](https://github.com/docker/docker-ce) repository.
